test/TestKrylov.f90

```diff
@@ -121,6 +123,99 @@ subroutine test_arnoldi_basis_orthogonality(error)
 
+   subroutine test_block_arnoldi_full_factorization(error)
+      ! This function checks the correctness of the block Arnoldi implementation by
+      ! verifying that the full factorization is correct, i.e. A @ X[1:p*k] = X[1:p*(k+1)] @ H.
+
+      !> Error type to be returned.
+      type(error_type), allocatable, intent(out) :: error
+      !> Test matrix.
+      class(rmatrix), allocatable :: A
+      !> Krylov subspace.
+      class(rvector), dimension(:), allocatable :: X
+      !> Krylov subspace dimension.
+      integer, parameter :: p = 2
+      integer, parameter :: kdim = test_size/p
+      !> Hessenberg matrix.
+      real(kind=wp) :: H(p*(kdim + 1), p*kdim)
+      !> Information flag.
+      integer :: info
+      !> Misc.
+      integer :: k
+      real(kind=wp) :: Xdata(test_size, p*(kdim + 1))
+      real(kind=wp) :: Rwrk(p,p)
+      double precision, dimension(p*kdim, p*kdim) :: G, Id
+      
+      ! --> Initialize matrix.
+      A = rmatrix(); call random_number(A%data)
+      ! --> Initialize Krylov subspace.
+      allocate (X(1:p*(kdim + 1))); 
+      do k = 1,p
+         call random_number(X(k)%data)
+      enddo
+      call qr_factorization(X(1:p), Rwrk, info)
+      do k = p+1, size(X)
+         call X(k)%zero()
+      end do
+      H = 0.0_wp
+
+      ! --> Arnoldi factorization.
+      call block_arnoldi_factorization(A, X, H, info, block_size = p) 
+      ! --> Check correctness of full factorization.
+      do k = 1, p*(kdim + 1)
+         Xdata(:, k) = X(k)%data
+      end do
+      call check(error, all_close(matmul(A%data, Xdata(:, 1:p*kdim)), matmul(Xdata, H), rtol, atol))
+
+      return
+   end subroutine test_block_arnoldi_full_factorization
+
+   subroutine test_block_arnoldi_basis_orthogonality(error)
+      !> Error type to be returned.
+      type(error_type), allocatable, intent(out) :: error
+      !> Test matrix.
+      class(rmatrix), allocatable :: A
+      !> Krylov subspace.
+      class(rvector), dimension(:), allocatable :: X
+      !> Krylov subspace dimension.
+      integer, parameter :: p = 2
+      integer, parameter :: kdim = test_size/p
+      !> Hessenberg matrix.
+      double precision, dimension(p*(kdim + 1), p*kdim) :: H
+      !> Information flag.
+      integer :: info
+      !> Misc.
+      double precision, dimension(p*kdim, p*kdim) :: G, Id
+      double precision :: Rwrk(p,p), alpha
+      integer :: i, j, k
+
+      !! --> Initialize random matrix.
+      A = rmatrix(); call random_number(A%data)
+      !! --> Initialize Krylov subspace.
+      allocate (X(1:p*(kdim + 1))); 
+      do k = 1,p
+         call random_number(X(k)%data)
+      enddo
+      call qr_factorization(X(1:p), Rwrk, info)
+      do k = p+1, size(X)
+         call X(k)%zero()
+      enddo
+      H = 0.0_wp
+      ! --> Arnoldi factorization.
+      call block_arnoldi_factorization(A, X, H, info, block_size = p)
+      ! --> Compute Gram matrix associated to the Krylov basis.
+      do i = 1, p*kdim
```

Comment:
Can't we now change these lines with one of the `mat_something` functions you implemented ?

Reply:
The same is true for many other places in the code. In fact, `qr_factorization` is essentially recursively calling `update_hessenberg_matrix` for each column. Within the hessenberg/QR functions, each loop 

```
do i = 1, k
  alpha = X(k + 1)%dot(X(i)); call X(k + 1)%axpby(1.0_wp, X(i), -alpha)
  H(i, k) = alpha
end do
```
can be written as
```
call mat_mult(H(1:k,k), X(1:k), X(k+1))
call mat_mult(Xwrk(1), X(1:k), H(1:k,k))
call mat_axpby(X(k+1), 1.0_wp, Xwrk(1), -1.0_wp)
```
but I'm not sure which version is "better". The advantage is, of course, that with the right indices the code for the single step is the same for the block case making the code redundant.

Ultimately, one could refactor the code base to include these functions. I think it's more readable than the loops, but that might be down to personal taste.